doc/neps/nep-0052-python-api-cleanup.rst

```diff
@@ -0,0 +1,331 @@
+
+
+Cleaning up the submodule structure
+```````````````````````````````````
+
+We will clean up the NumPy submodule structure, so it is easier to navigate.
+When this was discussed before (see
+`MAINT: Hide internals of np.lib to only show submodules <https://github.com/numpy/numpy/pull/18447>`__)
+there was already rough consensus on that - however it was hard to pull off in
+a minor release.
+
+A basic principle we will adhere to is "one function, one location". Functions
+that are exposed in more than one namespace (e.g., many functions are present
+in ``numpy`` and ``numpy.lib``) need to find a single home.
+
+We will reorganize the API reference guide along main and submodule namespaces,
+and only within the main namespace use the current subdivision along
+functionality groupings. Also by "mainstream" and special-purpose namespaces.
+Details TBD, something like:
+
+::
+
+    # Regular/recommended user-facing namespaces for general use. Present these
+    # as the primary set of namespaces to the users.
+    numpy
+    numpy.exceptions
+    numpy.fft
+    numpy.linalg
+    numpy.ma
+    numpy.polynomial
+    numpy.random
+    numpy.testing
+    numpy.typing
+
+    # Special-purpose namespaces. Keep these, but document them in a separate
+    # grouping in the reference guide and explain their purpose.
+    numpy.array_api
+    numpy.ctypeslib
+    numpy.emath
+    numpy.f2py  # only a couple of public functions, like `compile` and `get_include`
+    numpy.math
+    numpy.lib.stride_tricks
+    numpy.rec
+    numpy.types
+
+    # Legacy (prefer not to use). Third grouping in the reference guide.
+    numpy.char
+    numpy.distutils
+    numpy.matrixlib  # or deprecate?
+
+    # To remove
+    numpy.compat
+    numpy.core  # rename to _core
+    numpy.doc
+    numpy.matlib
+    numpy.version
+
+    # To clean out or somehow deal with: everything in `numpy.lib`
+
+.. note::
+
+    TBD: will we preserve ``np.lib`` or not? It only has a couple of unique
+    functions/objects, like ``Arrayterator`` (a candidate for removal) and the
+    ``stride_tricks`` subsubmodule. ``numpy.lib`` itself is not a coherent
+    namespace, and does not even have a reference guide page.
+
+.. note::
+
+   TBD: is ``numpy.math`` actually used at all? If not, should we hide/remove
+   it?
+
+We will make all submodules available lazily, so that users don't have to type
+``import numpy.xxx`` but can use ``import numpy as np; np.xxx.*``, while at the
+same time not negatively impacting the overhead of ``import numpy``. This has
+been very helpful for teaching scikit-image and SciPy.
+
+
+Reducing the number of ways to select dtypes
+````````````````````````````````````````````
+
+The many dtype classes, instances, aliases and ways to select them are one of
+the larger usability problems in the NumPy API. E.g.:
+
+.. code:: python
+
+   >>> # np.intp is different, but compares equal too
+   >>> np.int64 == np.int_ == np.dtype('i8') == np.sctypeDict['i8']
```

Comment:
I've always disliked this mixing up of a data type with a scalar variable class of that data type. If we can rid of those... (and perhaps have definitions for the standard dtypes, i.e., `f8 = np.dtype('f8')`? see below too)

Reply:
I'd like `__eq__` on dtypes to do the expected thing for Python objects indeed.

Less sure about `f8 = np.dtype('f8')` - I dislike the character codes a lot, they're too obscure. Scalars will hopefully go away completely at some point (they're not necessary for anything, and everyone seems to be in favor of them being removed eventually, it's just too big a lift for 2.0 it looks like).

Reply:
What I'd like is clear principles, like: 
- We have two sets of names for things:
  - the "bit-suffixed names" like `int32`, `float64` - with canonical names like `bool`, `void`, `object` added to that
  - the "C names" like `short`, `int`, `double`, `clongdouble`, etc.
- Other aliases that don't fall into these categories should either be removed, or de-emphasized in case we need them for backwards compatibility
- Avoid creating or using array scalars by instantiating them with `np.dtype` directly as much as possible, it's typically an anti-pattern

Reply:
Hmm, I'm super-used to using the string variants, but definitely can see your point. The annoyance is that `type(int32)` is not actually a `dtype` subclass. Anyway, that's for #23358; agreed here that a minimal set is a good idea.